### PR TITLE
Fix/dynamic chaininfo

### PR DESCRIPTION
**Pull request by tjandy98**
Currently, chain info is queried dynamically from `chainapsis/keplr-chain-registry`. However there are certain chains that are not registered, hence this chain info needs to be retrieved from `cosmos/chain-registry`.

### CODE REVIEW

File: src/constant/ibc.ts [hunk 1]
```diff
@@ -2159,3 +2159,107 @@ export type MinimalDenomMap = SimpleMap<string>;
...
+
+export interface CosmosChainsObj {
+  chains: ChainRegistryItem[];
+  repository: {
+    branch: string;
+    commit: string;
+    timestamp: string;
+    url: string;
+  };
+}
```
Comment (sarah-thong) [inline src/constant/ibc.ts:2265]: Nice work logging all the different types haha~ :)

File: src/modules/ibc.ts [hunk 2]
```diff
@@ -155,14 +164,116 @@ export class IBCModule extends BaseModule {
...
+
+    if (selectedChainData) {
+      try {
+        const chainInfoResponse = await fetch(`https://raw.githubusercontent.com/cosmos/chain-registry/master/${selectedChainData.chain_name}/chain.json`);
+        const chainInfoJson = await chainInfoResponse.json();
+        const chainAssetListResponse = await fetch(`https://raw.githubusercontent.com/cosmos/chain-registry/master/${selectedChainData.chain_name}/assetlist.json`);
+        const assetListJson = await chainAssetListResponse.json();
+        const features: string[] = [];
+      if (selectedChainData.cosmwasm_enabled) {
+        features.push("cosmwasm");
```
Comment (sarah-thong) [inline src/modules/ibc.ts:192]: Not sure if it's just me, but the indentation from this line onwards is a bit off. Please make sure the indentation is consistent.

File: src/modules/ibc.ts [hunk 3]
```diff
@@ -155,14 +164,116 @@ export class IBCModule extends BaseModule {
...
+      }
+
+      // Extract the staking currency denomination from chainInfoJson (min. denom)
+      const stakingCurrencyDenom = chainInfoJson['staking']['staking_tokens'][0]['denom'];
```
Comment (sarah-thong) [inline src/modules/ibc.ts:196]: 1. Actually, you don't need to use square brackets. Can use dot notation directly.
2. When you are using number as an index (i.e. for array), make sure to use optional chaining and have a fallback value.
i.e.
```typescript
const stakingCurrencyDenom = chainInfoJson.staking.staking_tokens[0]?.denom ?? '';
```

File: src/modules/ibc.ts [hunk 4]
```diff
@@ -155,14 +164,116 @@ export class IBCModule extends BaseModule {
...
+      // Find the highest denom unit
+      const maxExponentDenomUnit = stakingAsset.denom_units.reduce((prev: DenomUnit, current: DenomUnit) => {
+        return (prev.exponent > current.exponent) ? prev : current;
+      });
+
+      let stakeCurrency: Currency = {
+        coinDenom: maxExponentDenomUnit.denom.toUpperCase(),
+        coinMinimalDenom: stakingCurrencyDenom,
+        coinDecimals: maxExponentDenomUnit.exponent,
+        coinGeckoId: stakingAsset.coingecko_id,
```
Comment (sarah-thong) [inline src/modules/ibc.ts:212]: For stakingAsset, make sure to use optional chaining + provide a fallback value as Array.find() function is very likely to return null value.
```typescript
let stakeCurrency: Currency = {
  ...,
  coinGeckoId: stakingAsset?.coingecko_id ?? '',
}
```

File: src/modules/ibc.ts [hunk 5]
```diff
@@ -155,14 +164,116 @@ export class IBCModule extends BaseModule {
...
+          coinGeckoId: asset.coingecko_id,
+        };
+      });
+      
+
+      const feeCurrencies = chainInfoJson.fees.fee_tokens.map((feeToken: FeeToken) => {
+        // Find the corresponding asset for the fee token
+        const correspondingAsset = assetListJson.assets.find((asset: Asset) => asset.base === feeToken.denom);
+        if (!correspondingAsset) {
+          throw new Error(`No corresponding asset found for fee token denom: ${feeToken.denom}`);
```
Comment (sarah-thong) [inline src/modules/ibc.ts:234]: If you throw an error like this, the app/script using this function is likely to crash if the error is not caught properly. Maybe can revise this implementation such that if 
`!correspondingAsset`, the function skips this token.

File: src/modules/ibc.ts [hunk 6]
```diff
@@ -155,14 +164,116 @@ export class IBCModule extends BaseModule {
...
+            low: feeToken.low_gas_price,
+            average: feeToken.average_gas_price,
+            high: feeToken.high_gas_price,
+          }
+        };
+      });
+      
+        return {
+          rpc: chainInfoJson['apis']['rpc'][0]['address'],
+          rest: chainInfoJson['apis']['rest'][0]['address'],
```
Comment (sarah-thong) [inline src/modules/ibc.ts:255]: Same comment as line 196 of this file.

File: src/modules/ibc.ts (hunk 5 above)
Comment (tjandy98) [reply src/modules/ibc.ts]: @sarah-thong In this case would it be fine to use existing information instead?

```
return IBCUtils.EmbedChainInfos[chainId];
```

File: src/modules/ibc.ts (hunk 2 above)
Comment (tjandy98) [reply src/modules/ibc.ts]: Good catch, thanks!

File: src/modules/ibc.ts (hunk 3 above)
Comment (tjandy98) [reply src/modules/ibc.ts]: I've made the change, thank you!

File: src/modules/ibc.ts (hunk 5 above)
Comment (sarah-thong) [reply src/modules/ibc.ts]: This is not related to the issue. When you throw an error, the whole function fails with an error (i.e. you will not receive any output at all), which is a little stupid if it's only 1 or 2 asset infos that are missing.

As such, it's better to just continue the loop without adding the asset, thereby ensuring you have an output (however incomplete).

Comment (tjandy98) [reply src/modules/ibc.ts]: Thank you @sarah-thong. I have made the fix by filtering out fee tokens which do not have a corresponding asset. This way we will not run into the scenario of having undefined `correspondingAsset`

**Pull request merged**